File: crates/clawdstrike/src/engine.rs
//! HushEngine - Main entry point for security enforcement

use std::sync::Arc;
use tokio::sync::RwLock;
use tracing::{debug, info, warn};

use hush_core::receipt::{Provenance, Verdict, ViolationRef};
use hush_core::{sha256, Hash, Keypair, Receipt, SignedReceipt};
use serde::{Deserialize, Serialize};

use crate::async_guards::{AsyncGuard, AsyncGuardRuntime};
use crate::error::{Error, Result};
use crate::guards::{Guard, GuardAction, GuardContext, GuardResult, Severity};
use crate::policy::{Policy, PolicyGuards, RuleSet};

/// Per-guard evidence + an aggregated verdict.
#[must_use]
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct GuardReport {
    pub overall: GuardResult,
    pub per_guard: Vec<GuardResult>,
}

/// The main security enforcement engine
pub struct HushEngine {
    /// Active policy
    policy: Policy,
    /// Instantiated guards
    guards: PolicyGuards,
    /// Additional guards appended at runtime (evaluated after built-ins)
    extra_guards: Vec<Box<dyn Guard>>,
    /// Signing keypair (optional)
    keypair: Option<Keypair>,
    /// Session state
    state: Arc<RwLock<EngineState>>,
    /// Async guard runtime
    async_runtime: Arc<AsyncGuardRuntime>,
    /// Async guards instantiated from policy
    async_guards: Vec<Arc<dyn AsyncGuard>>,
    /// Async guard initialization error (fail closed)
    async_guard_init_error: Option<String>,
}

/// Engine session state
#[derive(Default)]
struct EngineState {
    /// Number of actions checked
    action_count: u64,
    /// Number of violations
    violation_count: u64,
    /// Recent violations
    violations: Vec<ViolationRef>,
}

impl HushEngine {
    /// Create a new engine with default policy
    pub fn new() -> Self {
        Self::with_policy(Policy::default())
    }

    /// Create with a specific policy
    pub fn with_policy(policy: Policy) -> Self {
        let guards = policy.create_guards();
        let async_runtime = Arc::new(AsyncGuardRuntime::new());
        let (async_guards, async_guard_init_error) =
            match crate::async_guards::registry::build_async_guards(&policy) {
                Ok(v) => (v, None),
                Err(e) => (Vec::new(), Some(e.to_string())),
            };

        Self {
            policy,
            guards,
            extra_guards: Vec::new(),
            keypair: None,
            state: Arc::new(RwLock::new(EngineState::default())),
            async_runtime,
            async_guards,
            async_guard_init_error,
        }
    }

    /// Create from a named ruleset
    pub fn from_ruleset(name: &str) -> Result<Self> {
        let ruleset = RuleSet::by_name(name)?
            .ok_or_else(|| Error::ConfigError(format!("Unknown ruleset: {}", name)))?;
        Ok(Self::with_policy(ruleset.policy))
    }

    /// Set the signing keypair
    pub fn with_keypair(mut self, keypair: Keypair) -> Self {
        self.keypair = Some(keypair);
        self
    }

    /// Generate a new signing keypair
    pub fn with_generated_keypair(mut self) -> Self {
        self.keypair = Some(Keypair::generate());
        self
    }

    /// Append an additional guard (evaluated after all built-in guards).
    ///
    /// Note: when `fail_fast` is enabled, guards after the first violation (including extras)
    /// will not run.
    pub fn with_extra_guard<G>(mut self, guard: G) -> Self
    where
        G: Guard + 'static,
    {
        self.extra_guards.push(Box::new(guard));
        self
    }

    /// Append an additional guard (evaluated after all built-in guards).
    ///
    /// Note: when `fail_fast` is enabled, guards after the first violation (including extras)
    /// will not run.
    pub fn with_extra_guard_box(mut self, guard: Box<dyn Guard>) -> Self {
        self.extra_guards.push(guard);
        self
    }

    /// Append an additional guard (evaluated after all built-in guards).
    ///
    /// Note: when `fail_fast` is enabled, guards after the first violation (including extras)
    /// will not run.
    pub fn add_extra_guard<G>(&mut self, guard: G) -> &mut Self
    where
        G: Guard + 'static,
    {
        self.extra_guards.push(Box::new(guard));
        self
    }

    /// Append an additional guard (evaluated after all built-in guards).
    ///
    /// Note: when `fail_fast` is enabled, guards after the first violation (including extras)
    /// will not run.
    pub fn add_extra_guard_box(&mut self, guard: Box<dyn Guard>) -> &mut Self {
        self.extra_guards.push(guard);
        self
    }

    /// Get the policy hash (derived from the policy YAML).
    ///
    /// Note: this does not include any runtime `extra_guards`.
    pub fn policy_hash(&self) -> Result<Hash> {
        let yaml = self.policy.to_yaml()?;
        Ok(sha256(yaml.as_bytes()))
    }

    /// Get the active policy.
    pub fn policy(&self) -> &Policy {
        &self.policy
    }

    /// Get the active policy YAML.
    pub fn policy_yaml(&self) -> Result<String> {
        self.policy.to_yaml()
    }

    /// Get the signing keypair, if configured.
    pub fn keypair(&self) -> Option<&Keypair> {
        self.keypair.as_ref()
    }

    /// Check a file access action
    pub async fn check_file_access(
        &self,
        path: &str,
        context: &GuardContext,
    ) -> Result<GuardResult> {
        self.check_action(&GuardAction::FileAccess(path), context)
            .await
    }

    /// Check a file write action
    pub async fn check_file_write(
        &self,
        path: &str,
        content: &[u8],
        context: &GuardContext,
    ) -> Result<GuardResult> {
        self.check_action(&GuardAction::FileWrite(path, content), context)
            .await
    }

    /// Check a network egress action
    pub async fn check_egress(
        &self,
        host: &str,
        port: u16,
        context: &GuardContext,
    ) -> Result<GuardResult> {
        self.check_action(&GuardAction::NetworkEgress(host, port), context)
            .await
    }

    /// Check a shell command action
    pub async fn check_shell(&self, command: &str, context: &GuardContext) -> Result<GuardResult> {
        self.check_action(&GuardAction::ShellCommand(command), context)
            .await
    }

    /// Check an MCP tool invocation
    pub async fn check_mcp_tool(
        &self,
        tool: &str,
        args: &serde_json::Value,
        context: &GuardContext,
    ) -> Result<GuardResult> {
        self.check_action(&GuardAction::McpTool(tool, args), context)
            .await
    }

    /// Check untrusted text (e.g. fetched web content) for prompt-injection signals.
    ///
    /// This uses `GuardAction::Custom("untrusted_text", ...)` and is evaluated by `PromptInjectionGuard`.
    pub async fn check_untrusted_text(
        &self,
        source: Option<&str>,
        text: &str,
        context: &GuardContext,
    ) -> Result<GuardResult> {
        let payload = match source {
            Some(source) => serde_json::json!({ "source": source, "text": text }),
            None => serde_json::json!({ "text": text }),
        };

        self.check_action(&GuardAction::Custom("untrusted_text", &payload), context)
            .await
    }

    /// Check a patch action
    pub async fn check_patch(
        &self,
        path: &str,
        diff: &str,
        context: &GuardContext,
    ) -> Result<GuardResult> {
        self.check_action(&GuardAction::Patch(path, diff), context)
            .await
    }

    /// Check any action against all applicable guards
    pub async fn check_action(
        &self,
        action: &GuardAction<'_>,
        context: &GuardContext,
    ) -> Result<GuardResult> {
        Ok(self.check_action_report(action, context).await?.overall)
    }

    /// Check any action and return per-guard evidence plus the aggregated verdict.
    pub async fn check_action_report(
        &self,
        action: &GuardAction<'_>,
        context: &GuardContext,
    ) -> Result<GuardReport> {
        if let Some(msg) = self.async_guard_init_error.as_ref() {
            return Err(Error::ConfigError(msg.clone()));
        }

        let builtins = self.guards.builtin_guards_in_order();
        let mut per_guard: Vec<GuardResult> =
            Vec::with_capacity(builtins.len() + self.extra_guards.len());

        for guard in builtins.chain(self.extra_guards.iter().map(|g| g.as_ref())) {
            if !guard.handles(action) {
                continue;
            }

            let result = guard.check(action, context).await;

            if self.policy.settings.effective_verbose_logging() {
                debug!(
                    guard = guard.name(),
                    allowed = result.allowed,
                    severity = ?result.severity,
                    "Guard check completed"
                );
            }

            if !result.allowed {
                let mut state = self.state.write().await;
                state.violation_count += 1;
                state.violations.push(ViolationRef {
                    guard: result.guard.clone(),
                    severity: format!("{:?}", result.severity),
                    message: result.message.clone(),
                    action: None,
                });

                warn!(
                    guard = result.guard,
                    message = result.message,
                    "Security violation detected"
                );
            }

            per_guard.push(result);

            if self.policy.settings.effective_fail_fast()
                && per_guard.last().is_some_and(|r| !r.allowed)
            {
                break;
            }
        }

        // If we've already denied locally, don't run async guards (avoids unnecessary network calls).
        if per_guard.iter().all(|r| r.allowed) && !self.async_guards.is_empty() {
            let async_results = self
                .async_runtime
                .evaluate_async_guards(
                    &self.async_guards,
                    action,
                    context,
                    self.policy.settings.effective_fail_fast(),
                )
                .await;

            for result in async_results {
                if self.policy.settings.effective_verbose_logging() {
                    debug!(
                        guard = result.guard,
                        allowed = result.allowed,
                        severity = ?result.severity,
                        "Guard check completed"
                    );
                }

                if !result.allowed {
                    let mut state = self.state.write().await;
                    state.violation_count += 1;
                    state.violations.push(ViolationRef {
                        guard: result.guard.clone(),
                        severity: format!("{:?}", result.severity),
                        message: result.message.clone(),
                        action: None,
                    });

                    warn!(
                        guard = result.guard,
                        message = result.message,
                        "Security violation detected"
                    );
                }

                per_guard.push(result);

                if self.policy.settings.effective_fail_fast()
                    && per_guard.last().is_some_and(|r| !r.allowed)
                {
                    break;
                }
            }
        }

        // Count the check even if we fail-fast.
        {
            let mut state = self.state.write().await;
            state.action_count += 1;
        }

        let overall = aggregate_overall(&per_guard);

        Ok(GuardReport { overall, per_guard })
    }

    /// Create a receipt for the current session
    pub async fn create_receipt(&self, content_hash: Hash) -> Result<Receipt> {
        let state = self.state.read().await;

        let verdict = if state.violation_count == 0 {
            Verdict::pass()
        } else {
            Verdict::fail()
        };

        let provenance = Provenance {
            clawdstrike_version: Some(env!("CARGO_PKG_VERSION").to_string()),
            provider: None,
            policy_hash: Some(self.policy_hash()?),
            ruleset: Some(self.policy.name.clone()),
            violations: state.violations.clone(),
        };

        let mut receipt = Receipt::new(content_hash, verdict).with_provenance(provenance);

        if !self.extra_guards.is_empty() {
            let extra_guards: Vec<&str> = self.extra_guards.iter().map(|g| g.name()).collect();
            receipt = receipt.with_metadata(serde_json::json!({
                "clawdstrike": {
                    "extra_guards": extra_guards,
                }
            }));
        }

        Ok(receipt)
    }

    /// Create and sign a receipt
    pub async fn create_signed_receipt(&self, content_hash: Hash) -> Result<SignedReceipt> {
        let keypair = self
            .keypair
            .as_ref()
            .ok_or_else(|| Error::ConfigError("No signing keypair configured".into()))?;

        let receipt = self.create_receipt(content_hash).await?;
        SignedReceipt::sign(receipt, keypair).map_err(Error::from)
    }

    /// Get session statistics
    pub async fn stats(&self) -> EngineStats {
        let state = self.state.read().await;
        EngineStats {
            action_count: state.action_count,
            violation_count: state.violation_count,
        }
    }

    /// Reset session state
    pub async fn reset(&self) {
        let mut state = self.state.write().await;
        *state = EngineState::default();
        info!("Engine state reset");
    }
}

impl Default for HushEngine {
    fn default() -> Self {
        Self::new()
    }
}

/// Session statistics
#[derive(Clone, Debug)]
pub struct EngineStats {
    pub action_count: u64,
    pub violation_count: u64,
}

/// Convert severity to ordinal for comparison
fn severity_ord(s: &Severity) -> u8 {
    match s {
        Severity::Info => 0,
        Severity::Warning => 1,
        Severity::Error => 2,
        Severity::Critical => 3,
    }
}

fn aggregate_overall(results: &[GuardResult]) -> GuardResult {
    if results.is_empty() {
        return GuardResult::allow("engine");
    }

    let mut best = &results[0];

    for r in &results[1..] {
        let best_blocks = !best.allowed;
        let r_blocks = !r.allowed;

        if r_blocks && !best_blocks {
            best = r;
            continue;
        }

        if r_blocks == best_blocks && severity_ord(&r.severity) > severity_ord(&best.severity) {
            best = r;
        }
    }

    best.clone()
}

#[cfg(test)]
mod tests {
    #![allow(clippy::expect_used, clippy::unwrap_used)]

    use super::*;
    use async_trait::async_trait;
    use std::sync::atomic::{AtomicUsize, Ordering};
    use std::sync::Arc;

    struct TestExtraGuard {
        name: &'static str,
        calls: Arc<AtomicUsize>,
    }

    #[async_trait]
    impl Guard for TestExtraGuard {
        fn name(&self) -> &str {
            self.name
        }

        fn handles(&self, action: &GuardAction<'_>) -> bool {
            match action {
                GuardAction::Custom(kind, _) => *kind == "extra_guard_test",
                GuardAction::FileAccess(_) => self.name == "extra_guard_order",
                _ => false,
            }
        }

        async fn check(&self, _action: &GuardAction<'_>, _context: &GuardContext) -> GuardResult {
            self.calls.fetch_add(1, Ordering::Relaxed);
            GuardResult::allow(self.name())
        }
    }

    #[tokio::test]
    async fn test_engine_new() {
        let engine = HushEngine::new();
        let stats = engine.stats().await;
        assert_eq!(stats.action_count, 0);
        assert_eq!(stats.violation_count, 0);
    }

    #[tokio::test]
    async fn test_check_file_access() {
        let engine = HushEngine::new();
        let context = GuardContext::new();

        // Normal file should be allowed
        let result = engine
            .check_file_access("/app/src/main.rs", &context)
            .await
            .unwrap();
        assert!(result.allowed);

        // SSH key should be blocked
        let result = engine
            .check_file_access("/home/user/.ssh/id_rsa", &context)
            .await
            .unwrap();
        assert!(!result.allowed);
    }

    #[tokio::test]
    async fn test_extra_guard_executes_for_custom_action() {
        let calls = Arc::new(AtomicUsize::new(0));

        let engine = HushEngine::new().with_extra_guard(TestExtraGuard {
            name: "extra_guard_test",
            calls: calls.clone(),
        });
        let context = GuardContext::new();
        let payload = serde_json::json!({ "test": true });

        let report = engine
            .check_action_report(&GuardAction::Custom("extra_guard_test", &payload), &context)
            .await
            .unwrap();

        assert_eq!(calls.load(Ordering::Relaxed), 1);
        assert_eq!(report.per_guard.len(), 1);
        assert_eq!(report.per_guard[0].guard, "extra_guard_test");
    }

    #[tokio::test]
    async fn test_extra_guard_runs_after_builtins() {
        let calls = Arc::new(AtomicUsize::new(0));

        let engine = HushEngine::new().with_extra_guard(TestExtraGuard {
            name: "extra_guard_order",
            calls: calls.clone(),
        });
        let context = GuardContext::new();

        let report = engine
            .check_action_report(&GuardAction::FileAccess("/app/src/main.rs"), &context)
            .await
            .unwrap();

        assert_eq!(calls.load(Ordering::Relaxed), 1);
        assert!(report.overall.allowed);
        assert!(report
            .per_guard
            .iter()
            .any(|r| r.guard != "extra_guard_order"));
        assert_eq!(
            report.per_guard.last().map(|r| r.guard.as_str()),
            Some("extra_guard_order")
        );
        assert_eq!(
            report
                .per_guard
                .iter()
                .filter(|r| r.guard == "extra_guard_order")
                .count(),
            1
        );
    }

    #[tokio::test]
    async fn test_fail_fast_skips_extra_guards_after_deny() {
        let calls = Arc::new(AtomicUsize::new(0));

        let mut policy = Policy::new();
        policy.settings.fail_fast = Some(true);

        let engine = HushEngine::with_policy(policy).with_extra_guard(TestExtraGuard {
            name: "extra_guard_order",
            calls: calls.clone(),
        });
        let context = GuardContext::new();

        let report = engine
            .check_action_report(&GuardAction::FileAccess("/home/user/.ssh/id_rsa"), &context)
            .await
            .unwrap();

        assert!(!report.overall.allowed);
        assert_eq!(calls.load(Ordering::Relaxed), 0);
    }

    #[tokio::test]
    async fn test_check_egress() {
        let engine = HushEngine::new();
        let context = GuardContext::new();

        // Allowed API
        let result = engine
            .check_egress("api.openai.com", 443, &context)
            .await
            .unwrap();
        assert!(result.allowed);

        // Unknown domain blocked
        let result = engine
            .check_egress("evil.com", 443, &context)
            .await
            .unwrap();
        assert!(!result.allowed);
    }

    #[tokio::test]
    async fn test_warn_aggregation_across_guards() {
        let engine = HushEngine::new();
        let context = GuardContext::new();

        let diff = r#"
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -1 +1 @@
+api_key = "0123456789abcdef0123456789abcdef"
"#;

        let report = engine
            .check_action_report(&GuardAction::Patch("src/lib.rs", diff), &context)
            .await
            .unwrap();

        assert!(report.overall.allowed);
        assert_eq!(report.overall.severity, Severity::Warning);
        assert!(report.per_guard.iter().any(|r| r.guard == "secret_leak"));
    }

    #[tokio::test]
    async fn test_violation_tracking() {
        let engine = HushEngine::new();
        let context = GuardContext::new();

        // Cause a violation
        let _ = engine
            .check_file_access("/home/user/.ssh/id_rsa", &context)
            .await
            .unwrap();

        let stats = engine.stats().await;
        assert_eq!(stats.action_count, 1);
        assert_eq!(stats.violation_count, 1);
    }

    #[tokio::test]
    async fn test_create_receipt() {
        let engine = HushEngine::new().with_generated_keypair();
        let context = GuardContext::new();

        // Normal action
        let _ = engine
            .check_file_access("/app/main.rs", &context)
            .await
            .unwrap();

        let content_hash = sha256(b"test content");
        let receipt = engine.create_receipt(content_hash).await.unwrap();

        assert!(receipt.verdict.passed);
        assert!(receipt.provenance.is_some());
    }

    #[tokio::test]
    async fn test_receipt_metadata_omitted_without_extra_guards() {
        let engine = HushEngine::new();
        let receipt = engine
            .create_receipt(sha256(b"test content"))
            .await
            .unwrap();
        assert!(receipt.metadata.is_none());
    }

    #[tokio::test]
    async fn test_receipt_metadata_includes_extra_guards() {
        let calls = Arc::new(AtomicUsize::new(0));

        let engine = HushEngine::new().with_extra_guard(TestExtraGuard {
            name: "extra_guard_metadata",
            calls: calls.clone(),
        });
        let receipt = engine
            .create_receipt(sha256(b"test content"))
            .await
            .unwrap();

        let metadata = receipt.metadata.expect("expected receipt metadata");
        assert_eq!(
            metadata["clawdstrike"]["extra_guards"],
            serde_json::json!(["extra_guard_metadata"])
        );
    }

    #[tokio::test]
    async fn test_create_signed_receipt() {
        let engine = HushEngine::new().with_generated_keypair();
        let context = GuardContext::new();

        let _ = engine
            .check_file_access("/app/main.rs", &context)
            .await
            .unwrap();

        let content_hash = sha256(b"test content");
        let signed = engine.create_signed_receipt(content_hash).await.unwrap();

        assert!(signed.receipt.verdict.passed);
    }

    #[tokio::test]
    async fn test_from_ruleset() {
        let engine = HushEngine::from_ruleset("strict").unwrap();
        let context = GuardContext::new();

        // Strict ruleset blocks unknown egress
        let result = engine
            .check_egress("random.com", 443, &context)
            .await
            .unwrap();
        assert!(!result.allowed);
    }

    #[tokio::test]
    async fn test_reset() {
        let engine = HushEngine::new();
        let context = GuardContext::new();

        let _ = engine
            .check_file_access("/home/user/.ssh/id_rsa", &context)
            .await
            .unwrap();
        assert_eq!(engine.stats().await.violation_count, 1);

        engine.reset().await;
        assert_eq!(engine.stats().await.violation_count, 0);
    }
}

File: packages/sdk/hush-go/engine/engine.go
// Package engine implements the HushEngine, the core orchestrator for
// Clawdstrike guard evaluation and receipt signing.
package engine

import (
	"errors"
	"fmt"
	"sync"
	"time"

	"github.com/backbay-labs/clawdstrike-go/crypto"
	"github.com/backbay-labs/clawdstrike-go/guards"
	"github.com/backbay-labs/clawdstrike-go/policy"
	"github.com/backbay-labs/clawdstrike-go/receipt"
)

// HushEngine orchestrates guards and signs receipts. Fail-closed: if a
// configuration error is stored, all subsequent checks return deny.
type HushEngine struct {
	mu sync.RWMutex

	keypair   *crypto.Keypair
	guards    []guards.Guard
	failFast  bool
	configErr error
	ruleset   string
}

// Builder constructs a HushEngine step by step.
type Builder struct {
	keypair    *crypto.Keypair
	guards     []guards.Guard
	failFast   bool
	configErr  error
	ruleset    string
}

func NewBuilder() *Builder {
	return &Builder{}
}

func (b *Builder) WithKeypair(kp *crypto.Keypair) *Builder {
	b.keypair = kp
	return b
}

func (b *Builder) WithGuard(g guards.Guard) *Builder {
	b.guards = append(b.guards, g)
	return b
}

func (b *Builder) WithFailFast(ff bool) *Builder {
	b.failFast = ff
	return b
}

func (b *Builder) WithRuleset(name string) *Builder {
	b.ruleset = name
	return b
}

// Build creates the HushEngine. Returns an error if configuration is invalid.
func (b *Builder) Build() (*HushEngine, error) {
	if b.configErr != nil {
		return nil, b.configErr
	}
	return &HushEngine{
		keypair:  b.keypair,
		guards:   b.guards,
		failFast: b.failFast,
		ruleset:  b.ruleset,
	}, nil
}

// FromRuleset creates a HushEngine from a named built-in ruleset.
// Guards are instantiated from the policy's guard configurations.
func FromRuleset(name string) (*HushEngine, error) {
	p, err := policy.ByName(name)
	if err != nil {
		return nil, err
	}
	return BuildFromPolicy(p)
}

// BuildFromPolicy creates a HushEngine from a resolved policy, instantiating
// all guards that have configuration entries.
func BuildFromPolicy(p *policy.Policy) (*HushEngine, error) {
	b := NewBuilder().WithRuleset(p.Name).WithFailFast(p.Settings.FailFast)

	if cfg := p.Guards.ForbiddenPath; cfg != nil && policy.GuardEnabled(cfg.Enabled) {
		b.WithGuard(guards.NewForbiddenPathGuard(cfg))
	}
	if cfg := p.Guards.EgressAllowlist; cfg != nil && policy.GuardEnabled(cfg.Enabled) {
		b.WithGuard(guards.NewEgressAllowlistGuard(cfg))
	}
	if cfg := p.Guards.SecretLeak; cfg != nil && policy.GuardEnabled(cfg.Enabled) {
		g, err := guards.NewSecretLeakGuard(cfg)
		if err != nil {
			return nil, fmt.Errorf("engine: instantiate secret_leak guard: %w", err)
		}
		b.WithGuard(g)
	}
	if cfg := p.Guards.PatchIntegrity; cfg != nil && policy.GuardEnabled(cfg.Enabled) {
		g, err := guards.NewPatchIntegrityGuard(cfg)
		if err != nil {
			return nil, fmt.Errorf("engine: instantiate patch_integrity guard: %w", err)
		}
		b.WithGuard(g)
	}
	if cfg := p.Guards.McpTool; cfg != nil && policy.GuardEnabled(cfg.Enabled) {
		b.WithGuard(guards.NewMcpToolGuard(cfg))
	}
	if cfg := p.Guards.PromptInjection; cfg != nil && policy.GuardEnabled(cfg.Enabled) {
		b.WithGuard(guards.NewPromptInjectionGuard(cfg))
	}
	if cfg := p.Guards.Jailbreak; cfg != nil && policy.GuardEnabled(cfg.Enabled) {
		b.WithGuard(guards.NewJailbreakGuard())
	}
	if cfg := p.Guards.SpiderSense; cfg != nil && policy.GuardEnabled(cfg.Enabled) {
		g, err := guards.NewSpiderSenseGuard(cfg)
		if err != nil {
			return nil, fmt.Errorf("engine: instantiate spider_sense guard: %w", err)
		}
		b.WithGuard(g)
	}

	return b.Build()
}

// SetConfigError marks the engine as misconfigured. All subsequent checks
// will return deny (fail-closed).
func (e *HushEngine) SetConfigError(err error) {
	e.mu.Lock()
	defer e.mu.Unlock()
	e.configErr = err
}

// AddGuard appends a guard to the engine pipeline.
func (e *HushEngine) AddGuard(g guards.Guard) {
	e.mu.Lock()
	defer e.mu.Unlock()
	e.guards = append(e.guards, g)
}

// Ruleset returns the configured ruleset name.
// Ruleset is safe for concurrent use; the field is immutable after Build().
func (e *HushEngine) Ruleset() string {
	return e.ruleset
}

func (e *HushEngine) Keypair() *crypto.Keypair {
	return e.keypair
}

// CheckAction evaluates the action against all guards and returns an
// aggregated GuardResult. Returns deny on configuration error (fail-closed).
func (e *HushEngine) CheckAction(action guards.GuardAction, ctx *guards.GuardContext) guards.GuardResult {
	e.mu.RLock()
	defer e.mu.RUnlock()

	// Fail-closed: config error → deny
	if e.configErr != nil {
		return guards.Block("engine", guards.Critical,
			fmt.Sprintf("engine configuration error: %v", e.configErr))
	}

	if ctx == nil {
		ctx = guards.NewContext()
	}

	var worstDeny *guards.GuardResult
	for _, g := range e.guards {
		if !g.Handles(action) {
			continue
		}
		result := g.Check(action, ctx)
		if !result.Allowed {
			if e.failFast {
				return result
			}
			if worstDeny == nil || result.Severity > worstDeny.Severity {
				r := result
				worstDeny = &r
			}
		}
	}

	if worstDeny != nil {
		return *worstDeny
	}
	return guards.Allow("engine")
}

// CheckActionReport evaluates the action and returns a detailed report
// with per-guard timing.
func (e *HushEngine) CheckActionReport(action guards.GuardAction, ctx *guards.GuardContext) *GuardReport {
	e.mu.RLock()
	defer e.mu.RUnlock()

	report := &GuardReport{Allowed: true}
	start := time.Now()

	// Fail-closed: config error → deny
	if e.configErr != nil {
		report.Allowed = false
		report.Results = append(report.Results, GuardResultEntry{
			GuardName: "engine",
			Result: guards.Block("engine", guards.Critical,
				fmt.Sprintf("engine configuration error: %v", e.configErr)),
			Duration: time.Since(start),
		})
		report.TotalDuration = time.Since(start)
		return report
	}

	if ctx == nil {
		ctx = guards.NewContext()
	}

	for _, g := range e.guards {
		if !g.Handles(action) {
			continue
		}
		guardStart := time.Now()
		result := g.Check(action, ctx)
		elapsed := time.Since(guardStart)

		report.Results = append(report.Results, GuardResultEntry{
			GuardName: g.Name(),
			Result:    result,
			Duration:  elapsed,
		})

		if !result.Allowed {
			report.Allowed = false
			if e.failFast {
				break
			}
		}
	}

	report.TotalDuration = time.Since(start)
	return report
}

// --- Convenience check methods ---

func (e *HushEngine) CheckFileAccess(path string) guards.GuardResult {
	return e.CheckAction(guards.FileAccess(path), nil)
}

func (e *HushEngine) CheckFileWrite(path string, content []byte) guards.GuardResult {
	return e.CheckAction(guards.FileWrite(path, content), nil)
}

func (e *HushEngine) CheckEgress(host string, port int) guards.GuardResult {
	return e.CheckAction(guards.NetworkEgress(host, port), nil)
}

func (e *HushEngine) CheckShell(cmd string) guards.GuardResult {
	return e.CheckAction(guards.ShellCommand(cmd), nil)
}

func (e *HushEngine) CheckMcpTool(name string, args interface{}) guards.GuardResult {
	return e.CheckAction(guards.McpTool(name, args), nil)
}

func (e *HushEngine) CheckPatch(file, diff string) guards.GuardResult {
	return e.CheckAction(guards.Patch(file, diff), nil)
}

func (e *HushEngine) CheckUntrustedText(text string) guards.GuardResult {
	return e.CheckAction(guards.Custom("untrusted_text", text), nil)
}

func (e *HushEngine) SignReceipt(r receipt.Receipt) (*receipt.SignedReceipt, error) {
	if e.keypair == nil {
		return nil, errors.New("engine has no signing keypair")
	}
	return receipt.Sign(r, e.keypair)
}
